application/src/main/java/org/opentripplanner/transit/model/filter/expr/ExpressionBuilder.java

```diff
@@ -22,7 +22,10 @@ public static <T> ExpressionBuilder<T> of() {
     return new ExpressionBuilder<>();
   }
 
-  public <V> ExpressionBuilder<T> or(Collection<V> values, Function<V, Matcher<T>> valueProvider) {
+  public <V> ExpressionBuilder<T> atLeastOneMatch(
+    Collection<V> values,
+    Function<V, Matcher<T>> valueProvider
+  ) {
     if (values.isEmpty()) {
       return this;
```

Comment:
After discussions yesterday I decided to go with creating a separate type for collections of criteria - the CriteriaCollection. There is currently one implementation of this, where the meaning of an empty list or a null list is that all filtered items are matched (which is the same as the behaviour in the API today), but if we want a different interpretation of what an empty or null list (or if there is a difference) means, then we can separate out an interface and create two implementations of it. Let me know what you think!
<!-- Reviewable comment -OB_bsYjEX_b0HKqgfbh:b-xe49fz -->
<!-- Sent from Reviewable.io -->

Reply:
I had a look at the `CriteriaCollection`. I think it has a bit of an issue, because it does a mix between encapsulating the logic for matching (`matchesAll()`) but still exposing all of it's internals (`values()`). And the user of this class has to know to first call `matchesAll()` and then do their own matching on the `values()`. This makes it pretty easy to make mistakes.

I think it would be better if the collection didn't know about the concept of "matching". Or if you want it to know about matching it should completely encapsulate the matching logic and only expose something like `boolean matches(T value)`. But that is probably not going to fit well since you already created the Matcher abstraction.

The nice thing about `Optional<List<E>>` (or similar) is that it's harder to make a mistake since it forces you to unwrap the optional before you can look at the values.

What do you think?

Reply:
> I had a look at the CriteriaCollection... This makes it pretty easy to make mistakes.

The `values()` should throw an exception if it matches all.

Reply:
The `CC` does give you an option of what empty means - the call can decide, if we use an Optonal, you need to put the logic in the input mapping. Here is how I would do it:

```Java
public abstract class CC<E> {
  @Nullable private Collection<E> criteria;
  
  private CC(@Nullable Collection<E> criteria) { this.criteria = criteria; }
  
  public static <E> CC<E> emptyMatchesAll(@Nullable Collection<E> criteria) {
    return new CC<>(criteria) {
      @Override public boolean matchesAll() { return criteria == null || criteria.isEmpty(); }
    };
  }
  public static <E> CC<E> emptyMatchesNothing(@Nullable Collection<E> criteria) {
    return new CC<>(criteria) {
      @Override public boolean matchesAll() { return criteria == null; }
    };
  }
  public abstract boolean matchesAll();
  public Collection<E> values() {
    if (matchesAll()) { throw new NoSuchElementException("..."); }
    return criteria;
  }
}

``` 
With this we have now documented the intentions in code - not relying on a fragile convention that the API knows how the consumer does it.

Reply:
> The `CC` does give you an option of what empty means - the caller can decide, if we use an Optonal, you need to put the logic in the input mapping.

In this case I think it would be good to put this logic in the input mapping since the collection is really only a data carrier from the api to the code that constructs the tree of `Matcher`s.

Another issue with having the collection know about matching by exposing a `matchesAll()` is that we can only use the collection for the whitelisting case where for example the user gives a list of agencies to allow. But we can't use the collection for blacklisting because then the default case would be to filter out everything.

I think we rather want to express the concept of `don't filter on this property` than `this matches on all values`.

Reply:
> Another issue with having the collection know about matching by exposing a matchesAll() is that we can only use the collection for whitelisting

This is not a problem, the `select` and  `not` is outside this. The `matchesAll()` is a contract with the filter logic, where the CC is defined. If we could, we should make it `package local`or a "friend" - but sadly it might not be possible in Java. 

> I think we rather want to express the concept of don't filter than match everything.

Not true, we want to express what an empty collection means and then inside the filter we need a way to query for that. We do not want to leave this error prune logic to the mapping layer of the API.

PS! The original idea is to make a type-safe optional. So, I think this is a matter of finding god names for the static factory-method `of()`, `values()` and `matchesAll()`. The `matchesAll()` is not so important since it is an internal contract of the filter component. But the factory methods are more important. 

What about the following factory methods:
 - `ofEmptyIsEverything(Collection<T> values) : EE<T>`
 - `ofEmptyIsNothing(Collection<T> values) : EE<T>`
 
And this internal methods: 
 - `includeEverything() : boolean`
 - `get() : Collection<T>` // Following the nameing convention of Java wrappers like `Optional` and `AtomicInteger`.

Reply:
Changed the names as you suggested @t2gran - @habrahamsson-skanetrafiken wdyt?

Reply:
> Is it a filter that matches anything?
No, the filter is defined as the composition of Matchers.
> Is it a collection containing all possible values?
That could be a way to think of it, but it's not the way we are going right now. We don't have any plans to interpret whether the caller has whitelisted all possible values.
> Is it a collection together with an interpretation of what empty should mean when building a matcher from this collection?
This is the goal currently (as I understand it) - force callers of the Transit API to be explicit about what an empty collection of criteria means when making calls into the Transit API (this is the reason for the ofEmptyIsEverything constructor - a potential future constructor would be ofEmptyIsNothing).

Forcing the callers to be explicit about what empty means when they construct the arguments for the Transit API reduces possible errors. Even with the helper method you suggest, the Request objects would take Optional<List<T>> and it would not be clear what the different configurations of Optional<List<T>> mean to the API.

I'll take a pass at improving the documentation of the CC and address your other comments. We'll discuss in the dev meeting later today too, I'm sure.